### DISCUSSION TITLE
PATH is not set correctly in nushell

### DISCUSSION THREAD

**Discussion by accidentaldevelopment** [jdx/mise#3555] [Troubleshooting and bug reports]
**Describe the bug**
This is technically two things, but both related to setting `PATH` in `nushell`. I have a fix locally, but am having some trouble getting the docker tasks to run, so opening an issue before I spend time on the tests.

If `mise` is not in the path when `mise activate nu` is run, it will prepend

```nushell
$env.PATH = ($env.PATH | prepend $YOUR_MISE_PATH)
```

to the output. This causes an error

```nushell
 1 │ $env.PATH = ($env.PATH | prepend '/opt/homebrew/bin')
   · ────┬────
   ·     ╰── expected def, const, extern, alias, use, module, export or export-env keyword
```

The solution is to use `export-env`:
```nushell
export-env { $env.PATH = ($env.PATH | prepend $YOUR_MISE_PATH) }
```

Related, the current method of setting `PATH` for installed tools breaks the table-like output expected in `nu`:

```nushell
> $env.PATH
$HOME/Desktop/mise/target/debug:$HOME/.local/share/mise/installs/neovim/nightly/bin:$HOME/.local/share/mise/installs/node/latest/bin:$HOME/.local/share/mise/installs/go/1.22.3/bin:$HOME/.local/share/mise/installs/go/1.22.3/go/bin:$HOME/.local/share/mise/installs/usage/latest/bin:/opt/homebrew/bin:/opt/homebrew/sbin:$HOME/.cargo/bin:/usr/bin:/bin:/usr/sbin:/sbin
```

instead of

```nushell
> $env.PATH
────┬─────────────────────────────────────────────────────
  0 │ $HOME/Desktop/mise/target/debug                     
  1 │ $HOME/.local/share/mise/installs/neovim/nightly/bin 
  2 │ $HOME/.local/share/mise/installs/node/latest/bin    
  3 │ $HOME/.local/share/mise/installs/go/1.22.3/bin      
  4 │ $HOME/.local/share/mise/installs/go/1.22.3/go/bin   
  5 │ $HOME/.local/share/mise/installs/usage/latest/bin   
  6 │ /opt/homebrew/bin                                   
  7 │ /opt/homebrew/sbin                                  
  8 │ $HOME/.cargo/bin                                    
  9 │ /usr/bin                                            
 10 │ /bin                                                
 11 │ /usr/sbin                                           
 12 │ /sbin                                               
────┴─────────────────────────────────────────────────────
```

**To Reproduce**
1. With `mise` not in PATH, run the setup instructions here: https://mise.jdx.dev/getting-started.html#nushell, but make sure to use the full path to `mise`.
2. Run the `use` command to load the module created above.

**Expected behavior**

**`mise doctor` output**

```text
version: 2024.6.6 macos-arm64 (2024-06-20)
activated: yes
shims_on_path: no

build_info: 
  Target: aarch64-apple-darwin
  Features: DEFAULT, NATIVE_TLS
  Built: Thu, 20 Jun 2024 17:07:53 +0000
  Rust Version: rustc 1.79.0 (129f3b996 2024-06-10) (Homebrew)
  Profile: release

shell: 
  nu
  0.94.2

dirs: 
  data: ~/.local/share/mise
  config: ~/.config/mise
  cache: ~/Library/Caches/mise
  state: ~/.local/state/mise
  shims: ~/.local/share/mise/shims

config_files: 
  ~/.config/mise/config.toml

backends: 
  cargo
  core
  go
  npm
  pipx
  spm
  ubi

plugins: 
  bun     (core)
  deno    (core)
  erlang  (core)
  go      (core)
  java    (core)
  neovim  https://github.com/richin13/asdf-neovim.git#d6118ad
  node    (core)
  python  (core)
  ruby    (core)
  usage   https://github.com/jdx/mise-usage.git#fe3888a
  zig     (core)

toolset: 
  neovim@nightly  
  node@22.2.0     
  go@1.22.3       
  usage@0.2.0     

env_vars: 
  MISE_SHELL=nu

settings: 
  activate_aggressive = false
  all_compile = false
  always_keep_download = false
  always_keep_install = false
  asdf_compat = true
  cargo_binstall = true
  color = true
  disable_default_shorthands = false
  disable_tools = []
  experimental = true
  go_default_packages_file = "~/.default-go-packages"
  go_download_mirror = "https://dl.google.com/go"
  go_repo = "https://github.com/golang/go"
  go_set_gopath = false
  go_set_goroot = true
  go_skip_checksum = false
  http_timeout = 30
  jobs = 4
  legacy_version_file = true
  legacy_version_file_disable_tools = []
  node_compile = false
  not_found_auto_install = true
  paranoid = false
  plugin_autoupdate_last_check_duration = "7d"
  python_default_packages_file = "/Users/username/.default-python-packages"
  python_pyenv_repo = "https://github.com/pyenv/pyenv.git"
  raw = false
  trusted_config_paths = []
  quiet = false
  verbose = false
  yes = false
  ci = false
  debug = false
  trace = false
  log_level = "info"
  python_venv_auto_create = false

  [status]
  missing_tools = "if_other_versions_installed"
  show_env = false
  show_tools = false

No warnings found
1 problem found:

1. shims are missing, run mise reshim to create them
   Missing shims: corepack, deno, go, gofmt, node, npm, npx, nvim, usage
```

**Additional context**
There is a relatively simple fix to both items:

```diff
diff --git a/src/shell/nushell.rs b/src/shell/nushell.rs
index 5a32641b..ed365c3c 100644
--- a/src/shell/nushell.rs
+++ b/src/shell/nushell.rs
@@ -67,7 +67,11 @@ impl Shell for Nushell {
           def --env "update-env" [] {{
             for $var in $in {{
               if $var.op == "set" {{
-                load-env {{($var.name): $var.value}}
+                if $var.name == 'PATH' {{
+                  $env.PATH = ($var.value | split row (char esep))
+                }} else {{
+                  load-env {{($var.name): $var.value}}
+                }}
               }} else if $var.op == "hide" {{
                 hide-env $var.name
               }}
@@ -97,7 +101,7 @@ impl Shell for Nushell {
     }
 
     fn prepend_env(&self, k: &str, v: &str) -> String {
-        format!("$env.{k} = ($env.{k} | prepend '{v}')\n")
+        format!("export-env {{ $env.{k} = ($env.{k} | prepend '{v}') }}\n")
     }
 
     fn unset_env(&self, k: &str) -> String {
```

Happy to open a PR if I can get the docker tasks to run.

**Comment by samuelallan72** [comment]
Please see https://github.com/jdx/mise/pull/3592 for a PR with a fix for setting PATH. :)

**Reply by Joniator** [reply]
This MR did not change the failing first line of the generated mise.nu. Im running nushell 0.106.1 and mise 2025.8.7 and get the same error, just with a different prepended path.

> ```nushell
>  1 │ $env.PATH = ($env.PATH | prepend '<my path prior to running mise activate>')
>    · ────┬────
>    ·     ╰── expected def, const, extern, alias, use, module, export or export-env keyword
> ```

I installed mise with pacman, in case the one line installer does some setup that the pacman script omits.

The script get's loaded at the end of my config.nu

```nushell
if (which mise | is-not-empty) {
    let mise_path = $autoload_dir | path join "mise.nu"
    ^mise activate nu | save --force $mise_path
}
```

However when using the script from the autoload, I get the info to trust the workspace I just changed into, so the hooks seem to have registered just fine. But I can replicate the problem that main gets registered as a command/alias for mise.

**Reply by samuelallan72** [reply]
> failing first line of the generated mise.nu. 

Ah nice thanks for the details, let me look into that error. I must have missed a case. I'll try to open a PR with a fix soon. :)

In the meantime, we should revert https://github.com/jdx/mise/pull/5981. The activation script (the generated mise.nu) is designed to be a nushell module (and it should remain so as a module should be more flexible, self contained, and re-usable). Adding it to autoload does appear to kinda work currently, as in it will update the PATH and register hooks, but then there are other issues: the mise wrapper command is registered as `main` instead, things like `mise deactivate` will no longer work, and it's just not designed to be used that way.

cc @jdx

**Reply by elken** [reply]
How funny @Joniator, I _just_ hit this myself and tried to find this to flag it 😅 

I get the same behaviour here though, putting the `$env.PATH` call inside the `export-env` works fine; for now I'm just using this script hardcoded

Related to the `nu` integration, I also get an error for the below:

```
 87 │ "__zoxide_hooked"="true"
    ·                   ───┬──
    ·                      ╰── can't convert string to boolean
```

If I call a nested `nu` inside `nu`

**Reply by samuelallan72** [reply]
@elken 

>  87 │ "__zoxide_hooked"="true"

This looks like an issue with zoxide - see https://github.com/ajeetdsouza/zoxide/issues/986

**Reply by samuelallan72** [reply]
I opened a PR that should hopefully fix this: https://github.com/jdx/mise/pull/6002 :)
